R/compile-dll.r

```diff
@@ -108,12 +108,33 @@ needs_compile <- function(pkg = ".") {
   source > dll
 }
 
+makevars <- function(pkg = ".") {
+  pkg <- as.package(pkg)
+  srcdir <- file.path(pkg$path, "src")
+  dir(srcdir, "^Makevars.*$", recursive = TRUE, full.names = TRUE)
+}
+
 # Does the package need a clean compile?
 # (i.e. is there a header or Makevars newer than the dll)
+# If the file Makevars includes a commented line like `# DEVTOOLS_NO_AUTO_CLEAN`,
+# then return FALSE and trust the Makevars to handle dependencies of
+# header files.
 needs_clean <- function(pkg = ".") {
   pkg <- as.package(pkg)
 
-  headers <- mtime(headers(pkg))
+  makevars <- makevars(pkg)
+  no_auto_clean <- FALSE
+  for (f in makevars) {
+    if (length(grep("^#+\\s*DEVTOOLS_NO_AUTO_CLEAN\\s*$", readLines(f, warn = FALSE))))
```

Comment:
Can you please extract out a `has_no_auto_clean()` function that takes a vector of paths, and returns TRUE or FALSE? (It's also slightly nicer to use `grepl()` rather than `grep()`)